core/drivers/stm32_gpio.c

```diff
@@ -77,6 +80,132 @@ static unsigned int gpio_lock;
 static STAILQ_HEAD(, stm32_gpio_bank) bank_list =
 		STAILQ_HEAD_INITIALIZER(bank_list);
 
+static struct stm32_gpio_bank *gpio_chip_to_bank(struct gpio_chip *chip)
+{
+	return container_of(chip, struct stm32_gpio_bank, gpio_chip);
+}
+
+static enum gpio_level stm32_gpio_get_level(struct gpio_chip *chip,
+					    unsigned int gpio_pin)
+{
+	struct stm32_gpio_bank *bank = gpio_chip_to_bank(chip);
+	enum gpio_level level = GPIO_LEVEL_HIGH;
+	unsigned int reg_offset = 0;
+	unsigned int mode = 0;
+
+	assert(gpio_pin < bank->ngpios);
+	clk_enable(bank->clock);
+
+	mode = (io_read32(bank->base + GPIO_MODER_OFFSET) >> (gpio_pin << 1)) &
+	       GPIO_MODE_MASK;
+
+	switch (mode) {
+	case GPIO_MODE_INPUT:
+		reg_offset = GPIO_IDR_OFFSET;
+		break;
+	case GPIO_MODE_OUTPUT:
+		reg_offset = GPIO_ODR_OFFSET;
+		break;
+	default:
+		panic();
+	}
+
+	if (io_read32(bank->base + reg_offset) & BIT(gpio_pin))
+		level = GPIO_LEVEL_HIGH;
+	else
+		level = GPIO_LEVEL_LOW;
+
+	clk_disable(bank->clock);
+
+	return level;
+}
+
+static void stm32_gpio_set_level(struct gpio_chip *chip, unsigned int gpio_pin,
+				 enum gpio_level level)
+{
+	struct stm32_gpio_bank *bank = gpio_chip_to_bank(chip);
+
+	assert(gpio_pin < bank->ngpios);
+	clk_enable(bank->clock);
+
+	if (level == GPIO_LEVEL_HIGH)
+		io_write32(bank->base + GPIO_BSRR_OFFSET, BIT(gpio_pin));
+	else
+		io_write32(bank->base + GPIO_BSRR_OFFSET, BIT(gpio_pin + 16));
+
+	clk_disable(bank->clock);
+}
+
+static enum gpio_dir stm32_gpio_get_direction(struct gpio_chip *chip,
+					      unsigned int gpio_pin)
+{
+	struct stm32_gpio_bank *bank = gpio_chip_to_bank(chip);
+	uint32_t mode = 0;
+
+	assert(gpio_pin < bank->ngpios);
+	clk_enable(bank->clock);
+
+	mode = (io_read32(bank->base + GPIO_MODER_OFFSET) >> (gpio_pin << 1)) &
```

Comment:
SHIFT_U32

Reply:
`gpio_pin` is an unsigned int, not a u32.